File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
         unordered_map<char, string> mp= {
                {'2', "abc"}, {'3', "def"}, {'4', "ghi"}, {'5', "jkl"}, {'6', "mno"}, {'7', "pqrs"}, {'8', "tuv"}, {'9', "wxyz"}
    };
    private:
     void findLetterCombination(int index, string &temp, vector<string> &result,string digits){
        if(index == digits.size()){
            result.push_back(temp);
            return;
        }

        string currString = mp[digits[index]];

        for(char c : currString){
            temp.push_back(c);
            findLetterCombination(index+1,temp,result,digits);
            temp.pop_back();
        }
     }


public:
    vector<string> letterCombinations(string digits) {
        string temp="";
        if(digits.size() == 0) return {};
        vector<string> result;

        findLetterCombination(0,temp,result,digits);
        return result;
    }
};

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.java
class Solution {
   HashMap<Character, String> mp = new HashMap<>();
   Solution(){
            mp.put('2', "abc");
            mp.put('3', "def");
            mp.put('4', "ghi");
            mp.put('5', "jkl");
            mp.put('6', "mno");
            mp.put('7', "pqrs");
            mp.put('8', "tuv");
            mp.put('9', "wxyz");
   }

    private void findLetters(int index, StringBuffer temp, List<String> result, String digits){
        if(index == digits.length()){
            result.add(temp.toString());
            return;
        }
        
        String current = mp.get(digits.charAt(index));
        for(char c : current.toCharArray()){
            temp.append(c);
            findLetters(index+1,temp,result,digits); // here index+1 not i+1;
            temp.deleteCharAt(temp.length()-1);
        }
    }
    public List<String> letterCombinations(String digits) {
       StringBuffer temp = new StringBuffer();
       List<String> result = new ArrayList<>();
       findLetters(0,temp,result,digits);
       return result;
    }
}